pyface/application.py

```diff
@@ -0,0 +1,259 @@
+from .i_window import IWindow
+
+logger = logging.getLogger(__name__)
+
+
+def clear_handlers(logger):
+    """ Remove all handlers from the given logger. """
+    for handler in logger.handlers[::-1]:
+        if hasattr(handler, 'close'):
+            handler.close()
+        logger.removeHandler(handler)
+
+
+class Application(BaseApplication):
+    """ A basic Pyface application.
+
+    This handles setting up starting up the GUI, window management, logging,
+    and other common features that we want when creating a GUI application.
+
+    Often you will want to subclass this application, but it can be used as-is
+    by hooking a listener to the `application_initialized` event::
+
+
+        class MainWindow(ApplicationWindow):
+            def _create_contents(self, parent):
+                self._label = HeadingText(parent, text="Hello World")
+                return self._label.control
+
+        def main():
+            application = Application(id='hello-world', name='Hello World')
+            window = MainWindow()
+            application.on_trait_change(
+                lambda: application.add_window(window),
+                'application_initialized'
+            )
+            application.run()
+
+    """
+
+    # -------------------------------------------------------------------------
+    # 'Application' interface
+    # -------------------------------------------------------------------------
+
+    # Branding ----------------------------------------------------------------
+
+    #: The splash screen for the application. No splash screen by default
+    splash_screen = Instance(ISplashScreen)
+
+    #: Icon for the application
+    icon = Instance(IImageResource)
+
+    # Window management -------------------------------------------------------
+
+    #: Default window size
+    window_size = Tuple((800, 600))
+
+    #: Currently active Window if any
+    active_window = Instance(IWindow)
+
+    #: List of all open windows in the application
+    windows = List(Instance(IWindow))
+
+    #: The Pyface GUI instance for the application
+    gui = Instance(IGUI)
+
+    # Window lifecycle methods -----------------------------------------------
+
+    def add_window(self, window):
+        """ Add a new window to the windows we are tracking. """
+        if window.size == (-1, -1):
+            window.size = self.window_size
+        if not window.title:
+            window.title = self.name
+        if self.icon:
+            window.icon = self.icon
+
+        # Keep a handle on all windows created so that non-active windows don't
+        # get garbage collected
+        self.windows.append(window)
+
+        # set up window listeners
+        # NOTE: 'activated' is not fired on a window when the window first
+        # opens and gets focus. It is only fired when the window comes from
+        # lower in the stack to be the active window.
+        window.on_trait_change(self._on_activate_window, 'activated')
```

Comment:
There should be no need to setup these change handlers here. The application should setup extended trait handlers looking into the `windows` list

e.g.
```
@on_trait_change('windows:activate'):
def _on_activate_window(self, object, name, value):
      window = object
      ...
```